exercises/nucleotide-count/example.hs

```diff
@@ -1,19 +1,18 @@
 module DNA (count, nucleotideCounts) where
-import Data.Map.Strict (Map, fromListWith)
 
-count :: Char -> String -> Int
-count needle = length . filter (nucleotide ==) . map verifyDNA
-  where nucleotide = verifyDNA needle
+import Data.Map.Strict (Map, (!), fromDistinctAscList)
+import Data.MultiSet   (fromList, occur)
 
-nucleotideCounts :: String -> Map Char Int
-nucleotideCounts strand = fromListWith (+) (defaults ++ map pair strand)
- where defaults = zip dna (repeat 0)
-       pair nucleotide = (verifyDNA nucleotide, 1)
+count :: Char -> String -> Either String Int
+count x xs = (!) <$> nucleotideCounts xs <*> valid x
 
-dna :: String
-dna = "ACGT"
+nucleotideCounts :: String -> Either String (Map Char Int)
+nucleotideCounts xs = fromDistinctAscList <$> mapM count' "ACGT"
+  where
+    count' x = (\c -> (x, c)) <$> occur' x
+    occur' x = occur x . fromList <$> mapM valid xs 
```

Comment:
All the uses of `<$>` would fail if `base < 4.8`, because it was only in `Control.Applicative` before that.

It's a little [tricky](https://ghc.haskell.org/trac/ghc/wiki/Migration/7.10#GHC7.10.xMigrationGuide) to use it in a compatible way, because if you `import Control.Applicative ((<$>))` it would raise a warning on newer versions, and we intentionally fail all tests with warnings on *Travis-CI*.